system-addon/common/PrerenderData.jsm

```diff
@@ -1,32 +1,80 @@
-const prefConfig = {
-  // Prefs listed with "invalidates: true" will prevent the prerendered version
+class _PrerenderData {
+  constructor(options) {
+    this.initialPrefs = options.initialPrefs;
+    this.initialSections = options.initialSections;
+    this._setValidation(options.validation);
+  }
+
+  get validation() {
+    return this._validation;
+  }
+
+  set validation(value) {
+    this._setValidation(value);
+  }
+
+  get invalidatingPrefs() {
+    return this._invalidatingPrefs;
+  }
+
+    // This is needed so we can use it in the constructor
+  _setValidation(value = []) {
+    this._validation = value;
+    this._invalidatingPrefs = value.reduce((result, next) => {
+      if (typeof next === "string") {
+        result.push(next);
+        return result;
+      } else if (next && next.oneOf) {
+        return result.concat(next.oneOf);
+      }
+      throw new Error("Your validation configuration is not properly configured");
+    }, []);
+  }
+
+  arePrefsValid(getPref) {
+    for (const name of this.validation) {
+      if (typeof name === "string") {
+        if (getPref(name) !== this.initialPrefs[name]) {
+          return false;
+        }
+      } else if (name && name.oneOf) {
+        let hasValid = false;
+        for (const innerName of name.oneOf) {
+          if (getPref(innerName) === this.initialPrefs[innerName]) {
```

Comment:
With every item being an array of either 1 item or multiple items, we could then just do
```js
for (const prefs of this.validation) {
  if (!prefs.some(name => getPref(name) === this.initialPrefs[name])) {
    return false;
  }
}
return true;
```